improver/lapse_rate.py

```diff
@@ -329,19 +332,85 @@ def _create_heightdiff_mask(self, all_orog_subsections):
+            self, temperature_data, orography_data, land_sea_mask_data,
+            all_temp_subsections, all_orog_subsections):
+        """
+        Calculate lapse rates and apply filters
+
+        Args:
+            temperature_data (numpy.ndarray)
+            orography_data (numpy.ndarray)
+            land_sea_mask_data (numpy.ndarray)
+            all_temp_subsections (numpy.ndarray)
+            all_orog_subsections (numpy.ndarray)
+
+        Returns:
+            numpy.ndarray
+                Lapse rate values
+        """
+        # Fill sea points with NaN values. Can't use Numpy mask since not
+        # recognised by "generic_filter" function.
+        temperature_data = np.where(
+            land_sea_mask_data, temperature_data, np.nan)
+
+        # Saves all neighbourhoods into "all_temp_subsections".
+        # cval is value given to points outside the array.
+        fnc = SaveNeighbourhood(allbuffers=all_temp_subsections)
+        generic_filter(temperature_data, fnc.filter,
+                       size=self.nbhood_size,
+                       mode='constant', cval=np.nan)
+
+        fnc = SaveNeighbourhood(allbuffers=all_orog_subsections)
+        generic_filter(orography_data, fnc.filter,
+                       size=self.nbhood_size,
+                       mode='constant', cval=np.nan)
+
+        # height_diff_mask is True for points where the height
+        # difference between the central point and its neighbours
+        # is > max_height_diff.
+        height_diff_mask = self._create_heightdiff_mask(
+            all_orog_subsections)
+
+        # Mask points with extreme height differences as NaN.
+        all_orog_subsections = np.where(height_diff_mask, np.nan,
+                                        all_orog_subsections)
+        all_temp_subsections = np.where(height_diff_mask, np.nan,
+                                        all_temp_subsections)
+
+        # Loop through both arrays and find gradient of each subsection.
+        # The gradient indicates lapse rate - save into another array.
+        # TODO: This for loop is the bottleneck in the code and needs to
+        # be parallelised.
+        lapse_rate_array = [self._calc_lapse_rate(temp, orog)
+                            for temp, orog in zip(all_temp_subsections,
+                                                  all_orog_subsections)]
+
+        lapse_rate_array = np.array(
+            lapse_rate_array, dtype=np.float32).reshape(temperature_data.shape)
+
+        # Enforce upper and lower limits on lapse rate values.
+        lapse_rate_array = np.where(lapse_rate_array < self.min_lapse_rate,
+                                    self.min_lapse_rate, lapse_rate_array)
+        lapse_rate_array = np.where(lapse_rate_array > self.max_lapse_rate,
+                                    self.max_lapse_rate, lapse_rate_array)
+        return lapse_rate_array
+
+    def process(self, temperature, orography, land_sea_mask,
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Too many local variables (24/20)](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=4906477)

Reply:
I'm going to ask for a stay of execution on this, given it relates to the aforementioned "horrible bit of code" that I'd like to redesign more widely in the imminent future.

Reply:
I should be clear: the particular horrible bit of code I was talking about is new, and my fault.  But I don't disagree in principle!